runner/scheduler.py

```diff
@@ -247,15 +247,36 @@ def pick_next_test_id(self, tests: dict[str, TestExample]) -> str:
             remaining_tokens = num_tokens - self.current_token_count
             while remaining_tokens > 0:
                 msg = filler_no_response_tokens_trivia(remaining_tokens, self.agent.max_message_size)
-                tokens_spent = self.send_message(SendMessageAction(msg, is_filling=True))
+                tokens_spent = self.send_message("", SendMessageAction(msg, is_filling=True))
                 remaining_tokens -= tokens_spent
             if not self.is_waiting(token_waiting_id, remove=True):
                 return token_waiting_id
 
         assert False, f"Couldn't find a test to run. Wait list: {self.wait_list}"
 
     def iter_tests(self, test_group: list[TestExample]) -> Iterator[TestExample]:
+        # Setup: get the list of tests that are in progress
+        test_actions_taken = self.master_log.get_tests_in_progress()
+        # TODO: More detailed assemblage of wait list
+        self.wait_list = {k: {"tokens": 0} for k in test_actions_taken.keys()}
         tests = {t.unique_id: t for t in test_group}
+
+        # Fast forward examples to the last action that was run
+        for run_id in self.wait_list.keys():
+            example = tests[run_id]
+            actions_to_ff = test_actions_taken[run_id]
+            for _ in range(actions_to_ff):
+                action = example.step()
+                if isinstance(action, SendAndRegisterAction):
+                    self.register_callback(example)
+
+                if isinstance(action, WaitAction) and action.time.seconds > 0:
+                    self.forward_time(seconds=action.time.seconds + 1)
+
+            # If the last action is a wait action, set the test to wait
+            if isinstance(action, WaitAction):
+                self.set_to_wait(run_id, action)
+
```

Comment:
Oh, I see... so these are the pre-recorded actions, and even if some of them depeded originally on the order (e.g. adding a quote to other test's message) it wouldn't matter, right? `run_id` being in the outermost loop somewhat confuses me.

Reply:
Precisely, `get_tests_in_progress()` gives us a dict of `{id: number of actions}` and this loop just moves the appropriate generators forward. We don't actually run the actions here unless they register a callback or change time.

Reply:
Right. What do you do with time changes? I guess that jumping to the latest time in the master log is enough, right?

Reply:
It probably would be sufficient, but we are more faithful than that. Each time change is repeated faithfully as on line 274.

Reply:
But this test has started from a point in time that is different from the original, so this will conflict with the timestamps that the agent has in its memory e.g. The run started at 1 pm and then a test moved 2h forward, so for the agent is 4 pm, but now you resume and it's 5 pm, and you move 2h forward to 7 pm.

Reply:
That ultimately doesn't matter as the timestamp dependent questions are dynamically generated. As long as the last timestamp in the log is in the past from the test's point of view, its fine. But we can just get the final element of the log and FF time to that point directly. It would make the FF loop a little cleaner.

Reply:
I was thinking from the agent's POV. It would be weird if my memories had timestamps from 5 pm, I think it's 9 pm, and I get asked about things that happened at 7 pm.